packages/design-system/src/components/Button/Button.tsx

```diff
@@ -54,144 +43,120 @@ type CommonButtonProps<T> = {
-    /**
-     * When provided the root component will render as an `<a>` element
-     * rather than `button`.
-     */
-    href?: string; // Still optional because it's optional on the anchor tag
-  };
-
-type ButtonButtonProps = CommonButtonProps<'button'> &
-  Omit<React.ComponentPropsWithRef<'button'>, OmitProps>;
-
-export type ButtonProps = CommonButtonProps<ButtonComponent> | LinkButtonProps | ButtonButtonProps;
-
-export class Button extends React.PureComponent<ButtonProps> {
-  static defaultProps = {
-    type: 'button',
-    component: 'button',
-  };
-
-  constructor(props: ButtonProps) {
-    super(props);
-
-    if (process.env.NODE_ENV !== 'production') {
-      if (props.inverse) {
-        console.warn(
-          `[Deprecated]: Please remove the 'inverse' prop in <Button>, use 'inversed' instead. This prop has been renamed and will be removed in a future release.`
-        );
-      }
-      if (props.variation === 'danger') {
-        console.warn(
-          `[Deprecated]: Please remove the 'danger' variation prop in <Button>. This prop has will be removed in a future release.`
-        );
-      }
-    }
+export type CustomButtonComponentType = React.ComponentType<any> | React.FC;
+export type ButtonComponentType = React.ElementType<any> | CustomButtonComponentType;
+export type NonAnchorButtonComponentType =
+  | CustomButtonComponentType
+  | React.ElementType<Exclude<string, 'a' | 'button'>>;
 
-    this.handleClick = this.handleClick.bind(this);
-    this.handleKeyPress = this.handleKeyPress.bind(this);
-  }
+type LinkTypeButtonProps = {
+  component?: 'a';
+  href: string;
+};
+type DefaultButtonTypeButtonProps = {
+  component?: 'button';
+  href?: undefined | null;
+};
+type OtherTypeButtonProps<T extends NonAnchorButtonComponentType> = {
+  component: T;
+  href?: undefined | null;
+};
 
-  // Get an object of props to pass to the rendered <Button> component
-  attrs(): any {
-    /**
-     * Since any number of arbitrary props can be passed into this component, we
-     * use a destructuring assignment to get only the props we want to pass to the
-     * rendered HTML element. For example, the "variation" prop is used to generate
-     * the classNames, but doesn't need to be passed to the rendered component, so
-     * we omit it here so that it's not included in the props object.
-     */
-    const {
-      className,
-      component,
-      inputRef,
-      inversed,
-      inverse,
-      onClick,
-      size,
-      variation,
-      ...props
-    } = this.props;
-
-    const attrs: any = {
-      className: this.classNames(),
-      ...props,
-    };
-
-    if (this.props.onClick) {
-      attrs.onClick = this.handleClick;
+export type ButtonProps<T extends ButtonComponentType> = PropsOf<T> &
+  CommonButtonProps &
+  (
+    | LinkTypeButtonProps
+    | DefaultButtonTypeButtonProps
+    | (T extends NonAnchorButtonComponentType ? OtherTypeButtonProps<T> : never)
+  );
```

Comment:
When coming across instances of `<Button>` in downstream apps, TypeScript wasn't able to determine which type of the intersection type `type ButtonProps = CommonButtonProps<ButtonComponent> | LinkButtonProps | ButtonButtonProps` was at play, so it would complain that props like `href` and `target` did not exist on `CommonButtonProps<ButtonComponent>` when it should have been using `LinkButtonProps`. The fix is to make the `Button` component itself a generic type, showing TypeScript which part of it is variable (`component`) and then try to help TypeScript narrow it down which specific type it is in this union (term is "discriminated union") by giving it hints about the `href` value and such.

The part about the last type in this union being `(T extends NonAnchorButtonComponentType ? OtherTypeButtonProps<T> : never)` is a later development, which I shall address in a follow-up comment.

Reply:
From my https://github.com/CMSgov/design-system/pull/1532/commits/ceeee3154e26b987814ce7ce9ceded4b6244a58f commit message:

The significance of this is that sometimes instances of Button even though they should be identified as having `LinkTypeButtonProps` (where `href` is strictly required)—like when explicitly passing `component='a'`—TypeScript seemed to be thinking it was the more generic `OtherTypeButtonProps<T>`. I _think_ this is working now.

Error that flags a missing `href` prop if the component has an explicit `component='a'` prop:
```
packages/design-system/src/components/Button/Button.test.tsx:36:8 - error TS2322: Type '{ component: "a"; target: "_blank"; type: string; children: string; }' is not assignable to type 'IntrinsicAttributes & ButtonProps<"a">'.
  Property 'href' is missing in type '{ component: "a"; target: "_blank"; type: string; children: string; }' but required in type 'LinkTypeButtonProps'.

36       <Button
          ~~~~~~

  packages/design-system/src/components/Button/Button.tsx:56:3
    56   href: string;
         ~~~~
    'href' is declared here.``
```

Note that it is correctly identifying this as a `LinkTypeButtonProps`.

Before this change, no error was thrown for a missing `href`, but in a similar situation where we explicitly pass a `component='a'` prop but give an _invalid_ value for `href`, we'd get the following error which illustrates TypeScript's incorrect inference about the type of the component props:
```
packages/design-system/src/components/Button/Button.test.tsx:36:8 - error TS2322: Type '{ component: "a"; href: { hello: string; }; target: "_blank"; type: string; children: string; }' is not assignable to type 'IntrinsicAttributes & ButtonProps<"a">'.
  Type '{ component: "a"; href: { hello: string; }; target: "_blank"; type: string; children: string; }' is not assignable to type 'OtherTypeButtonProps<"a">'.
    Types of property 'href' are incompatible.
      Type '{ hello: string; }' is not assignable to type 'null'.

36       <Button
          ~~~~~~
```
Note that it thinks our button has `OtherTypeButtonProps<"a">` props. We don't want that more generic type to be used when the more specific `LinkTypeButtonProps` should be used!!